### ISSUE TITLE
Add support for HTTP libraries (support PSR-7)

### ISSUE THREAD

**Issue by lanthaler** [lanthaler/JsonLD#4]
Currently `file_get_contents` is used to retrieve remote contexts and documents. This isn't really flexible and should sooner or later be replaced with a more sophisticated library.

Keep an eye on the [PHP FIG Proposal for a HTTP Client interface](https://github.com/php-fig/fig-standards/pull/24).

**Comment by lanthaler** [comment]
The [HTTP message interfaces (PSR-7)](http://www.php-fig.org/psr/psr-7/) have now been released. Update the processor to make use of it.

**Comment by cKlee** [comment]
To achieve this maybe http://httplug.io/ would be a good choice. As I understand the the [docs](http://docs.httplug.io/en/latest/) an own client implementation is no longer necessary. Meaning a separate class ```CurlLoad``` like you mentioned in https://github.com/lanthaler/JsonLD/pull/68 is also unnecessary but a HTTPlug [client-implementation](https://packagist.org/providers/php-http/client-implementation) will do this through composers ````require````. Is this the way you want to go?

**Comment by lanthaler** [comment]
Indeed, HTTPlug looks very promising. There seem to exist adapters for Guzzle and cURL but not file_get_contents (which often is the only thing that works on cheap, shared hosting). Too bad the PHP-FIG couldn’t agree on a client interface yet… but given how simple HTTPlug’s HttpClient interface is, I think this is the way to go for now. Do you have some time to work on this?

**Comment by cKlee** [comment]
I'm on holiday for two weeks now. After this I will look into it, if you have no other schedule. I'm just wonder why they say file_get_content is not important https://github.com/php-http/httplug/issues/32

**Comment by sagikazarmark** [comment]
Hey there,

HTTPlug developer here.

The fact that we marked it as not important doesn't mean you can't have an adapter for `file_get_contents`. We believe that it is not something which we can easily support and didn't really want to spend time with it. That said, if you would like to work on one, we will happily provide support if necessary.

Thanks for considering using HTTPlug.

**Comment by lanthaler** [comment]
Carsten, I doubt I’ll get to it before you. Would really appreciate if you could spend some cycles on it.

Thanks Márk for weighting in here. If we end up creating a file_get_contents adapter, where should it live? In one of our repos or directly under php-http?

**Comment by sagikazarmark** [comment]
@lanthaler it's up to you: if you want to contribute it to php-http, I can't see a reason why we couldn't host it, but it is not necessary to make it work. I guess we can add it to a third-party list if you decide to host it yourself.

**Comment by lanthaler** [comment]
OK, as it was raining, I had a couple of hours.. the result can be found at https://github.com/lanthaler/fgc-client :-)

The JsonLD library still needs to be refactored to use this. I have already started updating ml/iri to be PSR-7 conformant (https://github.com/lanthaler/IRI/tree/psr7)

**Comment by sagikazarmark** [comment]
You might also be interested in providing an UriFactory along with your IRI implementation.

**Comment by lanthaler** [comment]
Hmm… what depends on it? If it doesn’t provide a clear advantage, I would like to avoid adding a dependency on php-http/message-factory to ml/iri

**Comment by sagikazarmark** [comment]
The advantage: your package will work with HTTPlug out of the box. By adding additional Puli configuration, even the discovery layer will find your implementation and return the UriFactory to the user when an UriFactory is requested.

**Comment by lanthaler** [comment]
Understood. My question was what in HTTPlug depends on UriFactory. At the moment nothing seems to depend on it and thus I find it hard to justify pulling in an additional dependency.

**Comment by sagikazarmark** [comment]
I understand the reasons, I think this is the same reason why Guzzle rejected to do so.

However, this way your library (and guzzle psr7) are not really interoperable, or at least not switchable. There is no way to make a library using this package implementation independent, since the construction logic is unique for each implementation (unless you provide something which standardizes construction logic: factory).

I think this is enough justification (and given the fact that you already rely on another contract (psr7), I don't think adding another interface package is a problem).

This is my opinion of course. No hard feelings if you disagree. :smile:

**Comment by zobzn** [comment]
The first step should be to support injection of the loader interface.
https://github.com/lanthaler/JsonLD/pull/72
And then it is already possible to substitute any convenient interface implementation (FileGetContentsLoader, CurlLoader, HttpPugLoader, Guzzle6Loader, etc.)

**Comment by sagikazarmark** [comment]
Not sure if I understand: you want another abstraction layer above httplug?

**Comment by zobzn** [comment]
I want to be able to inject right now any other loader implementation instead of FileGetContentsLoader.
Maybe later it will be httpub, maybe.
But the most important thing is that now i can not use anything other than the built-in FileGetContentsLoader.

**Comment by tomgillett** [comment]
Following on from conversation in https://github.com/lanthaler/JsonLD/pull/100

> Absolutely. I started moving this into https://github.com/lanthaler/fgc-client a few years ago before those PSRs existed and used HTTPlug instead. If you are interested, we should update that package and then have JsonLD have depend on it and allow people to use other clients. The reason I like file_get_contents is that it supports both local and remote files and is available basically everywhere incl. the cheapest shared hosts.

I'd be happy to contribute some time to this. To explain what I'm looking to achieve...

I'm using this library within an application which makes integrates with lots of services via HTTP. If libraries support PSR-18 I can have my application provide a consistent client for all outbound HTTP requests. To me, this is better for maintenance and security and means I can add global configuration to that client if required (i.e. for an outbound proxy, for caching or for logging, etc.)

The alternative is an inconsisent mess - some libraries use `file_get_contents`, others cURL, others Guzzle - and I have little control over outbound traffic from my application.

I can already achieve this now by writing my own implementation of `ML\JsonLD\DocumentLoaderInterface` (see proof of concept here: https://github.com/tomgillett/JsonLD/commit/df1cff3b610a243cbba25988b25787e287a6b295) but native PSR-18 support feels cleaner to me.
